File: Sources/CodexStatusCore/ClaudeStatusMonitor.swift
import Foundation

public enum ClaudeState: String, Equatable {
    case idle
    case thinking
}

public struct ClaudeStatusSnapshot: Equatable {
    public let state: ClaudeState
    public let activeSessionCount: Int
    public let latestEventAt: Date?
    public let latestEventType: String?
    public let latestSessionFile: URL?
    public let latestSessionTitle: String?
    public let activeSessions: [StatusSessionSummary]
    public let idleSessions: [StatusSessionSummary]
    public let activeSessionTitles: [String]
    public let idleSessionTitles: [String]
    public let scannedFileCount: Int
    public let staleAfter: TimeInterval
    public let claudeHome: URL
    public let errorMessage: String?

    public var isThinking: Bool {
        state == .thinking
    }
}

public final class ClaudeStatusMonitor {
    public let claudeHome: URL
    public let staleAfter: TimeInterval

    private let projectsDirectory: URL
    private let fileManager: FileManager
    private let parser: ClaudeSessionActivityParser
    private let scanLimit: Int

    public init(
        claudeHome: URL = FileManager.default.homeDirectoryForCurrentUser.appendingPathComponent(".claude"),
        staleAfter: TimeInterval = 30 * 60,
        scanLimit: Int = 100,
        fileManager: FileManager = .default
    ) {
        self.claudeHome = claudeHome
        self.staleAfter = staleAfter
        self.scanLimit = scanLimit
        self.fileManager = fileManager
        self.projectsDirectory = claudeHome.appendingPathComponent("projects", isDirectory: true)
        self.parser = ClaudeSessionActivityParser()
    }

    public func snapshot(at now: Date = Date()) -> ClaudeStatusSnapshot {
        guard fileManager.fileExists(atPath: projectsDirectory.path) else {
            return ClaudeStatusSnapshot(
                state: .idle,
                activeSessionCount: 0,
                latestEventAt: nil,
                latestEventType: nil,
                latestSessionFile: nil,
                latestSessionTitle: nil,
                activeSessions: [],
                idleSessions: [],
                activeSessionTitles: [],
                idleSessionTitles: [],
                scannedFileCount: 0,
                staleAfter: staleAfter,
                claudeHome: claudeHome,
                errorMessage: "没有找到 Claude projects 目录"
            )
        }

        do {
            let files = try recentProjectFiles(limit: scanLimit)
            var latestActivity: ClaudeSessionActivity?
            var activeActivities: [ClaudeSessionActivity] = []
            var idleActivities: [ClaudeSessionActivity] = []

            for file in files {
                guard let activity = try parser.parse(url: file.url, modifiedAt: file.modifiedAt) else {
                    continue
                }

                if isNewer(activity, than: latestActivity) {
                    latestActivity = activity
                }

                if activity.isOpenTask, isFresh(activity, now: now) {
                    activeActivities.append(activity)
                } else {
                    idleActivities.append(activity)
                }
            }

            let representative = newestActivity(in: activeActivities) ?? latestActivity
            let activeSessions = sessionSummaries(from: activeActivities)
            let idleSessions = sessionSummaries(from: idleActivities)

            return ClaudeStatusSnapshot(
                state: activeActivities.isEmpty ? .idle : .thinking,
                activeSessionCount: activeActivities.count,
                latestEventAt: representative?.latestEventAt,
                latestEventType: representative?.latestEventType,
                latestSessionFile: representative?.fileURL,
                latestSessionTitle: representative.map(sessionTitle),
                activeSessions: activeSessions,
                idleSessions: idleSessions,
                activeSessionTitles: activeSessions.map(\.title),
                idleSessionTitles: idleSessions.map(\.title),
                scannedFileCount: files.count,
                staleAfter: staleAfter,
                claudeHome: claudeHome,
                errorMessage: nil
            )
        } catch {
            return ClaudeStatusSnapshot(
                state: .idle,
                activeSessionCount: 0,
                latestEventAt: nil,
                latestEventType: nil,
                latestSessionFile: nil,
                latestSessionTitle: nil,
                activeSessions: [],
                idleSessions: [],
                activeSessionTitles: [],
                idleSessionTitles: [],
                scannedFileCount: 0,
                staleAfter: staleAfter,
                claudeHome: claudeHome,
                errorMessage: error.localizedDescription
            )
        }
    }

    private func recentProjectFiles(limit: Int) throws -> [ClaudeProjectFile] {
        let resourceKeys: Set<URLResourceKey> = [.isRegularFileKey, .contentModificationDateKey]
        guard let enumerator = fileManager.enumerator(
            at: projectsDirectory,
            includingPropertiesForKeys: Array(resourceKeys),
            options: [.skipsHiddenFiles, .skipsPackageDescendants]
        ) else {
            return []
        }

        var files: [ClaudeProjectFile] = []
        for case let url as URL in enumerator where url.pathExtension == "jsonl" {
            let values = try? url.resourceValues(forKeys: resourceKeys)
            guard values?.isRegularFile == true, let modifiedAt = values?.contentModificationDate else {
                continue
            }
            files.append(ClaudeProjectFile(url: url, modifiedAt: modifiedAt))
        }

        return Array(files.sorted { $0.modifiedAt > $1.modifiedAt }.prefix(limit))
    }

    private func isFresh(_ activity: ClaudeSessionActivity, now: Date) -> Bool {
        let freshnessDate = max(activity.latestEventAt ?? activity.modifiedAt, activity.modifiedAt)
        return now.timeIntervalSince(freshnessDate) <= staleAfter
    }

    private func isNewer(_ activity: ClaudeSessionActivity, than other: ClaudeSessionActivity?) -> Bool {
        guard let other else {
            return true
        }
        return (activity.latestEventAt ?? activity.modifiedAt) > (other.latestEventAt ?? other.modifiedAt)
    }

    private func newestActivity(in activities: [ClaudeSessionActivity]) -> ClaudeSessionActivity? {
        activities.max { lhs, rhs in
            (lhs.latestEventAt ?? lhs.modifiedAt) < (rhs.latestEventAt ?? rhs.modifiedAt)
        }
    }

    private func sessionSummaries(from activities: [ClaudeSessionActivity]) -> [StatusSessionSummary] {
        activities
            .sorted {
                ($0.latestEventAt ?? $0.modifiedAt) > ($1.latestEventAt ?? $1.modifiedAt)
            }
            .map { activity in
                StatusSessionSummary(
                    id: activity.fileURL.path,
                    title: sessionTitle(for: activity),
                    lastAnswer: activity.lastAnswer
                )
            }
    }

    private func sessionTitle(for activity: ClaudeSessionActivity) -> String {
        SessionTitleNormalizer.displayTitle(activity.title)
    }
}

private struct ClaudeProjectFile {
    let url: URL
    let modifiedAt: Date
}

struct ClaudeSessionActivity {
    let fileURL: URL
    let modifiedAt: Date
    var title: String?
    var latestEventAt: Date?
    var latestEventType: String?
    var lastTaskStartedAt: Date?
    var lastTaskCompletedAt: Date?
    var lastAnswer: String?
    var sawAnyEvent = false

    var isOpenTask: Bool {
        if let lastTaskStartedAt, let lastTaskCompletedAt {
            return lastTaskStartedAt > lastTaskCompletedAt
        }

        return lastTaskStartedAt != nil
    }
}

final class ClaudeSessionActivityParser {
    private let tailByteLimit: UInt64
    private let isoWithFractionalSeconds: ISO8601DateFormatter
    private let isoWithoutFractionalSeconds: ISO8601DateFormatter

    init(tailByteLimit: UInt64 = 512 * 1024) {
        self.tailByteLimit = tailByteLimit
        self.isoWithFractionalSeconds = ISO8601DateFormatter()
        self.isoWithFractionalSeconds.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        self.isoWithoutFractionalSeconds = ISO8601DateFormatter()
        self.isoWithoutFractionalSeconds.formatOptions = [.withInternetDateTime]
    }

    func parse(url: URL, modifiedAt: Date) throws -> ClaudeSessionActivity? {
        let text = try tailText(from: url)
        var activity = ClaudeSessionActivity(fileURL: url, modifiedAt: modifiedAt)

        for line in text.split(separator: "\n") {
            guard let event = parseEvent(from: String(line)) else {
                continue
            }

            activity.sawAnyEvent = true
            activity.latestEventAt = event.timestamp
            activity.latestEventType = event.eventType
            if activity.title == nil {
                activity.title = event.titleCandidate
            }
            if let contentText = event.contentText {
                activity.lastAnswer = contentText
            }

            switch event.turnState {
            case .started:
                activity.lastTaskStartedAt = event.timestamp
            case .completed:
                activity.lastTaskCompletedAt = event.timestamp
            case .ignored:
                break
            }
        }

        return activity.sawAnyEvent ? activity : nil
    }

    private func tailText(from url: URL) throws -> String {
        let handle = try FileHandle(forReadingFrom: url)
        defer { try? handle.close() }

        let fileSize = try handle.seekToEnd()
        let bytesToRead = min(fileSize, tailByteLimit)
        let offset = fileSize - bytesToRead
        try handle.seek(toOffset: offset)

        let data = try handle.readToEnd() ?? Data()
        var text = String(decoding: data, as: UTF8.self)

        if offset > 0, let firstNewline = text.firstIndex(of: "\n") {
            text.removeSubrange(text.startIndex...firstNewline)
        }

        return text
    }

    private func parseEvent(from line: String) -> ParsedClaudeEvent? {
        guard let data = line.data(using: .utf8),
              let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let timestampText = object["timestamp"] as? String,
              let timestamp = parseTimestamp(timestampText),
              let topLevelType = object["type"] as? String
        else {
            return nil
        }

        let message = object["message"] as? [String: Any]
        let role = message?["role"] as? String
        let stopReason = message?["stop_reason"] as? String
        let titleCandidate = SessionTitleNormalizer.explicitTitle(in: object)
            ?? userTitleCandidate(topLevelType: topLevelType, message: message)
        let contentText = topLevelType == "assistant"
            ? SessionTitleNormalizer.title(fromContent: message?["content"], maxLength: 200)
            : nil

        switch topLevelType {
        case "user":
            return ParsedClaudeEvent(
                timestamp: timestamp,
                eventType: "user",
                contentText: nil,
                turnState: .started,
                titleCandidate: titleCandidate
            )
        case "assistant":
            if stopReason == "end_turn" {
                return ParsedClaudeEvent(
                    timestamp: timestamp,
                    eventType: "assistant:end_turn",
                    contentText: contentText,
                    turnState: .completed,
                    titleCandidate: titleCandidate
                )
            }

            let suffix = stopReason ?? role ?? "assistant"
            return ParsedClaudeEvent(
                timestamp: timestamp,
                eventType: "assistant:\(suffix)",
                contentText: contentText,
                turnState: .started,
                titleCandidate: titleCandidate
            )
        default:
            return ParsedClaudeEvent(
                timestamp: timestamp,
                eventType: topLevelType,
                contentText: nil,
                turnState: .ignored,
                titleCandidate: titleCandidate
            )
        }
    }

    private func parseTimestamp(_ timestamp: String) -> Date? {
        isoWithFractionalSeconds.date(from: timestamp) ?? isoWithoutFractionalSeconds.date(from: timestamp)
    }

    private func userTitleCandidate(topLevelType: String, message: [String: Any]?) -> String? {
        guard topLevelType == "user" || message?["role"] as? String == "user" else {
            return nil
        }

        return SessionTitleNormalizer.title(fromContent: message?["content"])
    }
}

private enum ClaudeTurnState {
    case started
    case completed
    case ignored
}

private struct ParsedClaudeEvent {
    let timestamp: Date
    let eventType: String
    let contentText: String?
    let turnState: ClaudeTurnState
    let titleCandidate: String?
}

File: win/OneMenu.Core/Monitors/ClaudeStatusMonitor.cs
using System.Text.Json;
using OneMenu.Core.Infrastructure;
using OneMenu.Core.Models;

namespace OneMenu.Core.Monitors;

public class ClaudeStatusMonitor
{
    private readonly string _claudeHome;
    private readonly TimeSpan _staleAfter;
    private readonly string _projectsDirectory;
    private readonly int _scanLimit;
    private readonly ClaudeJsonParser _parser;

    public string ClaudeHome => _claudeHome;

    public ClaudeStatusMonitor(
        string? claudeHome = null,
        TimeSpan? staleAfter = null,
        int scanLimit = 100)
    {
        _claudeHome = claudeHome
            ?? Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".claude");
        _staleAfter = staleAfter ?? TimeSpan.FromMinutes(30);
        _scanLimit = scanLimit;
        _projectsDirectory = Path.Combine(_claudeHome, "projects");
        _parser = new ClaudeJsonParser();
    }

    public ClaudeStatusSnapshot Snapshot(DateTime? now = null)
    {
        var nowValue = now ?? DateTime.UtcNow;

        if (!Directory.Exists(_projectsDirectory))
        {
            return ClaudeStatusSnapshot.Empty(_staleAfter, _claudeHome, "没有找到 Claude projects 目录");
        }

        try
        {
            var files = RecentProjectFiles(_scanLimit);
            ClaudeSessionActivity? latestActivity = null;
            var activeActivities = new List<ClaudeSessionActivity>();
            var idleActivities = new List<ClaudeSessionActivity>();

            foreach (var file in files)
            {
                var activity = _parser.Parse(file.path, file.modifiedAt);
                if (activity == null) continue;

                if (IsNewer(activity, latestActivity))
                    latestActivity = activity;

                if (activity.IsOpenTask && IsFresh(activity, nowValue))
                    activeActivities.Add(activity);
                else
                    idleActivities.Add(activity);
            }

            var representative = NewestActivity(activeActivities) ?? latestActivity;
            var activeSessions = SessionSummaries(activeActivities);
            var idleSessions = SessionSummaries(idleActivities);

            return new ClaudeStatusSnapshot(
                State: activeActivities.Count == 0 ? ClaudeState.Idle : ClaudeState.Thinking,
                ActiveSessionCount: activeActivities.Count,
                LatestEventAt: representative?.LatestEventAt,
                LatestEventType: representative?.LatestEventType,
                LatestSessionFile: representative?.FilePath,
                LatestSessionTitle: representative != null ? SessionTitle(representative) : null,
                ActiveSessions: activeSessions,
                IdleSessions: idleSessions,
                ActiveSessionTitles: activeSessions.Select(s => s.Title).ToArray(),
                IdleSessionTitles: idleSessions.Select(s => s.Title).ToArray(),
                ScannedFileCount: files.Count,
                StaleAfter: _staleAfter,
                ClaudeHome: _claudeHome,
                ErrorMessage: null);
        }
        catch (Exception ex)
        {
            return ClaudeStatusSnapshot.Empty(_staleAfter, _claudeHome, ex.Message);
        }
    }

    private List<(string path, DateTime modifiedAt)> RecentProjectFiles(int limit)
    {
        var files = new List<(string path, DateTime modifiedAt)>();

        try
        {
            foreach (var filePath in Directory.EnumerateFiles(_projectsDirectory, "*.jsonl",
                SearchOption.AllDirectories))
            {
                try
                {
                    var fileInfo = new FileInfo(filePath);
                    files.Add((filePath, fileInfo.LastWriteTimeUtc));
                }
                catch
                {
                    // skip inaccessible files
                }
            }
        }
        catch
        {
            // directory enumeration failed
        }

        return files
            .OrderByDescending(f => f.modifiedAt)
            .Take(limit)
            .ToList();
    }

    private bool IsFresh(ClaudeSessionActivity activity, DateTime now)
    {
        var freshnessDate = Max(activity.LatestEventAt ?? activity.ModifiedAt, activity.ModifiedAt);
        return (now - freshnessDate) <= _staleAfter;
    }

    private static bool IsNewer(ClaudeSessionActivity activity, ClaudeSessionActivity? other)
    {
        if (other == null) return true;
        return (activity.LatestEventAt ?? activity.ModifiedAt) >
               (other.LatestEventAt ?? other.ModifiedAt);
    }

    private static ClaudeSessionActivity? NewestActivity(List<ClaudeSessionActivity> activities)
    {
        return activities.MaxBy(a => a.LatestEventAt ?? a.ModifiedAt);
    }

    private static List<StatusSessionSummary> SessionSummaries(List<ClaudeSessionActivity> activities)
    {
        return activities
            .OrderByDescending(a => a.LatestEventAt ?? a.ModifiedAt)
            .Select(a => new StatusSessionSummary(
                Id: a.FilePath,
                Title: SessionTitle(a),
                LastAnswer: a.LastAnswer))
            .ToList();
    }

    private static string SessionTitle(ClaudeSessionActivity activity) =>
        SessionTitleNormalizer.DisplayTitle(activity.Title);

    private static DateTime Max(DateTime a, DateTime b) => a > b ? a : b;

    /// <summary>
    /// JSONL event parser for Claude session files.
    /// </summary>
    private class ClaudeJsonParser
    {
        private static readonly string[] IsoFormats =
        {
            "yyyy-MM-dd'T'HH:mm:ss.FFFFFFF'Z'",
            "yyyy-MM-dd'T'HH:mm:ss.FFFFFFFzzz",
            "yyyy-MM-dd'T'HH:mm:ss'Z'",
            "yyyy-MM-dd'T'HH:mm:sszzz"
        };

        private enum TurnState { Started, Completed, Ignored }

        public ClaudeSessionActivity? Parse(string filePath, DateTime modifiedAt)
        {
            try
            {
                var text = new JsonlFileReader().ReadTail(filePath);
                var activity = new ClaudeSessionActivity
                {
                    FilePath = filePath,
                    ModifiedAt = modifiedAt
                };

                foreach (var line in text.Split('\n'))
                {
                    var trimmed = line.Trim();
                    if (trimmed.Length == 0) continue;

                    var evt = ParseEvent(trimmed);
                    if (evt == null) continue;

                    activity.SawAnyEvent = true;
                    activity.LatestEventAt = evt.Value.Timestamp;
                    activity.LatestEventType = evt.Value.EventType;
                    activity.Title ??= evt.Value.TitleCandidate;
                    if (evt.Value.ContentText != null)
                        activity.LastAnswer = evt.Value.ContentText;

                    switch (evt.Value.TurnState)
                    {
                        case TurnState.Started:
                            activity.LastTaskStartedAt = evt.Value.Timestamp;
                            break;
                        case TurnState.Completed:
                            activity.LastTaskCompletedAt = evt.Value.Timestamp;
                            break;
                    }
                }

                return activity.SawAnyEvent ? activity : null;
            }
            catch
            {
                return null;
            }
        }

        private (DateTime Timestamp, string EventType, string? TitleCandidate, string? ContentText,
            TurnState TurnState)? ParseEvent(string line)
        {
            try
            {
                using var doc = JsonDocument.Parse(line);
                var root = doc.RootElement;

                if (!root.TryGetProperty("timestamp", out var tsEl) ||
                    tsEl.ValueKind != JsonValueKind.String)
                    return null;
                if (!TryParseTimestamp(tsEl.GetString()!, out var timestamp))
                    return null;
                if (!root.TryGetProperty("type", out var typeEl) ||
                    typeEl.ValueKind != JsonValueKind.String)
                    return null;

                var topLevelType = typeEl.GetString()!;

                JsonElement? message = root.TryGetProperty("message", out var msg) &&
                    msg.ValueKind == JsonValueKind.Object ? msg : null;

                var role = message?.TryGetProperty("role", out var r) == true &&
                    r.ValueKind == JsonValueKind.String ? r.GetString() : null;
                var stopReason = message?.TryGetProperty("stop_reason", out var sr) == true &&
                    sr.ValueKind == JsonValueKind.String ? sr.GetString() : null;

                var titleCandidate = SessionTitleNormalizer.ExplicitTitleIn(root)
                    ?? UserTitleCandidate(topLevelType, message);

                string? contentText = topLevelType == "assistant"
                    ? SessionTitleNormalizer.TitleFromContent(
                        message?.TryGetProperty("content", out var c) == true ? c : null, 200)
                    : null;

                return topLevelType switch
                {
                    "user" => (timestamp, "user", titleCandidate, null, TurnState.Started),
                    "assistant" when stopReason == "end_turn" =>
                        (timestamp, "assistant:end_turn", titleCandidate, contentText, TurnState.Completed),
                    "assistant" =>
                        (timestamp, $"assistant:{stopReason ?? role ?? "assistant"}", titleCandidate,
                         contentText, TurnState.Started),
                    _ => (timestamp, topLevelType, titleCandidate, null, TurnState.Ignored)
                };
            }
            catch
            {
                return null;
            }
        }

        private static string? UserTitleCandidate(string topLevelType, JsonElement? message)
        {
            if (topLevelType != "user" && (message?.TryGetProperty("role", out var r) != true ||
                r.ValueKind != JsonValueKind.String || r.GetString() != "user"))
                return null;

            if (message?.TryGetProperty("content", out var content) == true)
                return SessionTitleNormalizer.TitleFromContent(content);
            return null;
        }

        private static bool TryParseTimestamp(string text, out DateTime result)
        {
            if (DateTime.TryParseExact(text, IsoFormats,
                    System.Globalization.CultureInfo.InvariantCulture,
                    System.Globalization.DateTimeStyles.AssumeUniversal |
                    System.Globalization.DateTimeStyles.AdjustToUniversal,
                    out result))
                return true;

            if (DateTime.TryParse(text, null,
                    System.Globalization.DateTimeStyles.RoundtripKind, out result))
            {
                result = result.ToUniversalTime();
                return true;
            }

            return false;
        }
    }
}
